### ISSUE TITLE
Double window controls in some CSD applications

### ISSUE THREAD

**Issue by lah7** [lah7/gtk3-classic#32]
### Bug Description
`gnome-disk-utility 40.0` shows double window controls: the actual window controls and a duplicate within the application. There is also a black border on the right/bottom.

The latest `gtk3-classic` against the older `3.38.2` does not inhibit the problem, so the GNOME developers made a change in their code.

Thankfully, it is still a GTK3 application and not GTK4, despite the versioning.

### Expected Behaviour
The window controls inside the application should be hidden.

### Workaround

Downgrade and pin the older version.

### Screenshots

40.0
![gnome-disks-40](https://user-images.githubusercontent.com/13032135/111920679-32f47900-8a88-11eb-813d-e6146d5a64e2.png)

3.38.2
![gnome-disks-3 38 2](https://user-images.githubusercontent.com/13032135/111920681-36880000-8a88-11eb-9eeb-d3866f7d0caa.png)

### Environment
 - OS: EndeavourOS
 - Desktop Environment: KDE
 - gtk3-classic Version: 3.24.27

**Comment by lah7** [comment]
Likewise, `simple-scan` 40.0 inhibits the same problem: double controls, and black border.

## 40.0
![Screenshot_20210326_114039](https://user-images.githubusercontent.com/13032135/112627231-3b063d00-8e29-11eb-8e77-d1197fb68e5c.png)
![Screenshot_20210326_114150](https://user-images.githubusercontent.com/13032135/112627234-3b9ed380-8e29-11eb-8555-d361d0d911c8.png)


## 3.38.5
![Screenshot_20210326_114407](https://user-images.githubusercontent.com/13032135/112627238-3e99c400-8e29-11eb-9bd1-01fded201d92.png)
![Screenshot_20210326_114413](https://user-images.githubusercontent.com/13032135/112627240-3e99c400-8e29-11eb-9c8b-628ffcc4b718.png)

**Comment by MR-Diamond** [comment]
Also gnome-calculator 40.0

**Comment by wb9688** [comment]
My ***guess*** would be that those are the apps using libhandy's `HdyWindow` and `HdyHeaderBar` and that that causes this behavior.

**Comment by ZaWertun** [comment]
Just fixed similar bug in my fork of gtk3-nocsd: https://github.com/ZaWertun/gtk3-nocsd/tree/test/libhandy.
Some screenshots:
![изображение](https://user-images.githubusercontent.com/1080298/122485025-a43bbd80-cfde-11eb-840a-9f39b71d398b.png)
![изображение](https://user-images.githubusercontent.com/1080298/122485057-b3bb0680-cfde-11eb-98e6-365979429447.png)

**Comment by simifor** [comment]
I looked at @ZaWertun's last commit, then browsed a bit gtk's tree to see where it'd and made a patch 
[patch.txt](https://github.com/lah7/gtk3-classic/files/6888632/patch.txt) full credits to zawertun really. This only fixes the black outlines seen in some places, as I didn't even touch the CSD related commits.

**Comment by lah7** [comment]
Thank you @simifor, your `patch.txt` made it very easy to incorporate into the quilt patch system. Committed as 12f90e216bd894b27cfedc5b1eb918fd7cae4811

Thank you to @ZaWertun as well. I'm a bit lost on which commits to incorporate by looking at `gtk3-nocsd`'s repository. Any help is much appreciated if anyone wishes to create a patch for this repository to fix this _regression_. I get confused with `quilt` too. 😅

**Comment by N0rbert** [comment]
Great! Thanks for merging.
I think, we need to skip/disable both `csd__clean-headerbar.patch` and `csd__disabled-by-default.patch` temporarily for Ubuntu MATE 21.10 *impish* [deb-packages](https://launchpad.net/~lah7/+archive/ubuntu/gtk3-classic?field.series_filter=impish) to get GNOME native apps look better here, so without double borders.

**Comment by lah7** [comment]
I noticed CSD applications under patched Ubuntu 21.10 doesn't look great, depending on the theme:

![Screenshot_20220103_183047](https://user-images.githubusercontent.com/13032135/147966422-8872d415-01ae-43d3-9306-05d050db8f4d.png)
![Screenshot_20220103_183118](https://user-images.githubusercontent.com/13032135/147966424-e13b1e40-05d8-40bb-a501-0e0ff3ac2497.png)
![Screenshot_20220103_183138](https://user-images.githubusercontent.com/13032135/147966425-a72c42e2-ae5a-46b0-a4b1-1496533b7452.png)

For Ubuntu 21.10 users, I will revert those patches in the packaging since they ship newer (40.x) GNOME apps. Just until the "new" CSD is fixed, pull requests much appreciated!

**Edit:** In 22.04, this returns to using the default patchset, so double controls are expected.

**Comment by daemonspudguy** [comment]
I also noticed this with the program Pamac on Arch Linux.

**Comment by servimo** [comment]
I have this with EasyEffects on Manjaro too.

**Comment by ghost** [comment]
The same bug happens on gedit.
![image](https://user-images.githubusercontent.com/89735533/175778702-7274bdec-3d9b-48e3-a7f2-5ae6568b9795.png)

**Comment by fenglengshun** [comment]
It seems to not be all, as I noticed that Lutris and FSearch (when CSD is enabled) doesn't have that issue while Komikku and Pamac has them. Potentially because of some hardcoded settings in their CSD setup?

<details><summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/26621899/183274026-b1e08276-b279-45de-8722-c924365f33af.png)

![image](https://user-images.githubusercontent.com/26621899/183274035-bad2760f-bcad-41a1-b18f-8b73124bfe2b.png)

![image](https://user-images.githubusercontent.com/26621899/183274007-fc09748c-b35f-4385-b45b-24022285ac21.png)

![image](https://user-images.githubusercontent.com/26621899/183274014-43bcfa8f-b78d-492d-b1f2-0bcedcf5b328.png)

</details>

For now the only thing that can be done is to just use `GTK_CSD=` set to `0` or `1`. Which isn't an issue with those two apps I use, as I don't need the KDE window decoration features often with them and it's just an eyesore when not maximized.

Edit: For now the other thing that can be done is by manually changing `gtk-decoration-layout=` to `:menu,appmenu` in `~/.config/gtk-3.0/settings.ini` though I don't know if this would be fine outside of KDE or not.

**Comment by ghost** [comment]
TL;DR: Application upstream design issue (not GTK)

I can't speak to all the examples given, but just analyzing the OP application (gnome-disk-utility):
- Upstream [commit](https://github.com/GNOME/gnome-disk-utility/commit/922e0c1daea2778c5ff14546b0912ae9b30b6eda) that dropped off support for a SSD bar
- Subsequent later commits switched over to using libhandy (a CSD UI library that operates on top of GTK)

I imagine (as most of the other examples are GNOME apps) that they did a similar switch to libhandy. In this case there's little we can do with a patched GTK to solve this issue (as libhandy is building the title bar and has no such checks to skip doing so on SSD or classic-style distributions).

**Comment by probonopd** [comment]
Opened an upstream issue:
https://gitlab.gnome.org/GNOME/libhandy/-/issues/470

Please consider upvoting that upstream issue if you agree that this should be fixed in libhandy.

**Comment by lah7** [comment]
I think the next course of action is to have a separately patched  [`libhandy`](https://archlinux.org/packages/extra/x86_64/libhandy/) package that removes the CSD window buttons.

Personally I'm not up for maintaining that, but I'll happily link if anyone comes up with an AUR package for the offending library.

**Comment by probonopd** [comment]
Upstream [says](https://gitlab.gnome.org/GNOME/libhandy/-/issues/470):

> Even if I was interested in supporting that thing (I'm not), there's no way to do it from libhandy in the first place as it has no idea about any patches your GTK may or may not have. This isn't something that's possible to "support". It's like asking a website to support a user style better.

At least we have asked.

**Comment by ra50** [comment]
> I think the next course of action is to have a separately patched [`libhandy`](https://archlinux.org/packages/extra/x86_64/libhandy/) package that removes the CSD window buttons.
> 
> Personally I'm not up for maintaining that, but I'll happily link if anyone comes up with an AUR package for the offending library.

Can look into this.  Maybe it's as simple as setting `priv->shows_wm_decorations = FALSE;` in `hdy_header_bar_set_show_close_button`? (here https://gitlab.gnome.org/GNOME/libhandy/-/blob/main/src/hdy-header-bar.c)
